Don't build criterion in CI

It would be nice if criterion was not a regular dev-depenency that is built for tests since we only need it for benchmarks that we don't want to run on CI either way.

The one way to do this I can come up with so far is moving benchmarks to a separate, workspace-excluded crate.

This seems like a bit of a rough edge of rust features, enabling a file global feature then makes it impossible to do anything with `#[cfg(not(same_as_global))]`.

This is the error now
```
error[E0601]: `main` function not found in crate `event_deserialize`
   --> ruma-events/benches/event_deserialize.rs:6:1
    |
6   | / #![cfg(feature = "criterion_bench")]
7   | |
8   | | use criterion::{criterion_group, criterion_main, Criterion};
9   | | use ruma_events::{
...   |
110 | | // ENABLE THE "criterion_bench" FEATURE TO RUN BENCHMARKS
111 | | criterion_main!(benches);
    | |_________________________^ consider adding a `main` function to `ruma-events/benches/event_deserialize.rs`
```
It's not great but it is only the one error and maybe with an all caps comment it would be visible

```
error: Enable the criterion_bench feature to run benchmarks
   --> ruma-events/benches/event_deserialize.rs:120:1
    |
120 | compile_error!("Enable the criterion_bench feature to run benchmarks");
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0601]: `main` function not found in crate `event_deserialize`
   --> ruma-events/benches/event_deserialize.rs:6:1
    |
6   | / #![allow(unused_imports)]
7   | |
8   | | #[cfg(feature = "criterion_bench")]
9   | | use criterion::{criterion_group, criterion_main, Criterion};
...   |
119 | | #[cfg(not(feature = "criterion_bench"))]
120 | | compile_error!("Enable the criterion_bench feature to run benchmarks");
    | | ^ consider adding a `main` function to `ruma-events/benches/event_deserialize.rs`
```

--- a/ruma-events/Cargo.toml
+++ b/ruma-events/Cargo.toml
@@ -34,7 +34,6 @@ optional = true
 
 [features]
 default = []
-criterion_bench = ["criterion"]
 
 [[bench]]
 name = "event_deserialize"

--- a/ruma-events/Cargo.toml
+++ b/ruma-events/Cargo.toml
@@ -13,6 +13,7 @@ version = "0.21.3"
 edition = "2018"
 
 [dependencies]
+criterion = { version = "0.3.2", optional = true }
 js_int = { version = "0.1.7", features = ["serde"] }
 ruma-common = { version = "0.1.3", path = "../ruma-common" }
 ruma-events-macros = { version = "=0.21.3", path = "../ruma-events-macros" }
@@ -27,13 +28,6 @@ maplit = "1.0.2"
 matches = "0.1.8"
 ruma-identifiers = { version = "0.16.2", path = "../ruma-identifiers", features = ["rand"] }
 trybuild = "1.0.30"
-
-[dependencies.criterion]
-version = "0.3.2"
-optional = true
-
-[features]
-default = []
 
 [[bench]]
 name = "event_deserialize"

--- a/ruma-events/Cargo.toml
+++ b/ruma-events/Cargo.toml
@@ -27,7 +27,14 @@ maplit = "1.0.2"
 matches = "0.1.8"
 ruma-identifiers = { version = "0.16.2", path = "../ruma-identifiers", features = ["rand"] }
 trybuild = "1.0.30"
-criterion = "0.3.2"
+
+[dependencies.criterion]
+version = "0.3.2"
+optional = true
+
+[features]
+default = []
+criterion_bench = ["criterion"]
 
 [[bench]]
 name = "event_deserialize"

--- a/ruma-events/benches/event_deserialize.rs
+++ b/ruma-events/benches/event_deserialize.rs
@@ -2,11 +2,11 @@
 // or pass any other args to it, it fails with the error
 // `cargo bench unknown option --save-baseline`.
 // To pass args to criterion, use this form
-// `cargo bench --features criterion_bench --bench <name of the bench> -- --save-baseline <name>`.
+// `cargo bench --features criterion --bench <name of the bench> -- --save-baseline <name>`.
 
 #![allow(unused_imports, dead_code)]
 
-#[cfg(feature = "criterion_bench")]
+#[cfg(feature = "criterion")]
 use criterion::{criterion_group, criterion_main, Criterion};
 use ruma_events::{
     room::power_levels::PowerLevelsEventContent, AnyEvent, AnyRoomEvent, AnyStateEvent, EventJson,
@@ -47,7 +47,7 @@ fn power_levels() -> serde_json::Value {
     })
 }
 
-#[cfg(feature = "criterion_bench")]
+#[cfg(feature = "criterion")]
 fn deserialize_any_event(c: &mut Criterion) {
     let json_data = power_levels();
 
@@ -61,7 +61,7 @@ fn deserialize_any_event(c: &mut Criterion) {
     });
 }
 
-#[cfg(feature = "criterion_bench")]
+#[cfg(feature = "criterion")]
 fn deserialize_any_room_event(c: &mut Criterion) {
     let json_data = power_levels();
 
@@ -75,7 +75,7 @@ fn deserialize_any_room_event(c: &mut Criterion) {
     });
 }
 
-#[cfg(feature = "criterion_bench")]
+#[cfg(feature = "criterion")]
 fn deserialize_any_state_event(c: &mut Criterion) {
     let json_data = power_levels();
 
@@ -89,7 +89,7 @@ fn deserialize_any_state_event(c: &mut Criterion) {
     });
 }
 
-#[cfg(feature = "criterion_bench")]
+#[cfg(feature = "criterion")]
 fn deserialize_specific_event(c: &mut Criterion) {
     let json_data = power_levels();
 
@@ -105,7 +105,7 @@ fn deserialize_specific_event(c: &mut Criterion) {
     });
 }
 
-#[cfg(feature = "criterion_bench")]
+#[cfg(feature = "criterion")]
 criterion_group!(
     benches,
     deserialize_any_event,
@@ -114,12 +114,12 @@ criterion_group!(
     deserialize_specific_event
 );
 
-#[cfg(feature = "criterion_bench")]
+#[cfg(feature = "criterion")]
 criterion_main!(benches);
 
-#[cfg(not(feature = "criterion_bench"))]
+#[cfg(not(feature = "criterion"))]
 fn main() {
-    // To run the benchmarks the "criterion_bench" feature must be enabled.
-    // `cargo bench --features criterion_bench --bench <name of the bench>`
-    panic!("Enable the criterion_bench feature to run benchmarks");
+    // To run the benchmarks the "criterion" feature must be enabled use:
+    // `cargo bench --features criterion --bench event_deserialize`
+    panic!("Enable the criterion feature to run benchmarks");
 }

--- a/ruma-events/benches/event_deserialize.rs
+++ b/ruma-events/benches/event_deserialize.rs
@@ -2,8 +2,16 @@
 // or pass any other args to it, it fails with the error
 // `cargo bench unknown option --save-baseline`.
 // To pass args to criterion, use this form
-// `cargo bench --bench <name of the bench> -- --save-baseline <name>`.
+// `cargo bench --features criterion_bench --bench <name of the bench> -- --save-baseline <name>`.
 
+#![allow(unused_imports, dead_code)]
+
+// To run the benchmarks the "criterion_bench" feature must be enabled.
+// `cargo bench --features criterion_bench --bench <name of the bench>`
+#[cfg(not(feature = "criterion_bench"))]
+compile_error!("Enable the criterion_bench feature to run benchmarks");
+
+#[cfg(feature = "criterion_bench")]
 use criterion::{criterion_group, criterion_main, Criterion};
 use ruma_events::{
     room::power_levels::PowerLevelsEventContent, AnyEvent, AnyRoomEvent, AnyStateEvent, EventJson,
@@ -44,6 +52,7 @@ fn power_levels() -> serde_json::Value {
     })
 }
 
+#[cfg(feature = "criterion_bench")]
 fn deserialize_any_event(c: &mut Criterion) {
     let json_data = power_levels();
 
@@ -57,6 +66,7 @@ fn deserialize_any_event(c: &mut Criterion) {
     });
 }
 
+#[cfg(feature = "criterion_bench")]
 fn deserialize_any_room_event(c: &mut Criterion) {
     let json_data = power_levels();
 
@@ -70,6 +80,7 @@ fn deserialize_any_room_event(c: &mut Criterion) {
     });
 }
 
+#[cfg(feature = "criterion_bench")]
 fn deserialize_any_state_event(c: &mut Criterion) {
     let json_data = power_levels();
 
@@ -83,6 +94,7 @@ fn deserialize_any_state_event(c: &mut Criterion) {
     });
 }
 
+#[cfg(feature = "criterion_bench")]
 fn deserialize_specific_event(c: &mut Criterion) {
     let json_data = power_levels();
 
@@ -98,6 +110,7 @@ fn deserialize_specific_event(c: &mut Criterion) {
     });
 }
 
+#[cfg(feature = "criterion_bench")]
 criterion_group!(
     benches,
     deserialize_any_event,
@@ -106,4 +119,10 @@ criterion_group!(
     deserialize_specific_event
 );
 
+#[cfg(feature = "criterion_bench")]
 criterion_main!(benches);
+
+// This makes the compiler only spit out the `compile_error!` when
+// `cargo bench` is run without the `criterion_bench` feature.
+#[cfg(not(feature = "criterion_bench"))]
+fn main() {}

--- a/ruma-events/benches/event_deserialize.rs
+++ b/ruma-events/benches/event_deserialize.rs
@@ -6,11 +6,6 @@
 
 #![allow(unused_imports, dead_code)]
 
-// To run the benchmarks the "criterion_bench" feature must be enabled.
-// `cargo bench --features criterion_bench --bench <name of the bench>`
-#[cfg(not(feature = "criterion_bench"))]
-compile_error!("Enable the criterion_bench feature to run benchmarks");
-
 #[cfg(feature = "criterion_bench")]
 use criterion::{criterion_group, criterion_main, Criterion};
 use ruma_events::{
@@ -122,7 +117,9 @@ criterion_group!(
 #[cfg(feature = "criterion_bench")]
 criterion_main!(benches);
 
-// This makes the compiler only spit out the `compile_error!` when
-// `cargo bench` is run without the `criterion_bench` feature.
 #[cfg(not(feature = "criterion_bench"))]
-fn main() {}
+fn main() {
+    // To run the benchmarks the "criterion_bench" feature must be enabled.
+    // `cargo bench --features criterion_bench --bench <name of the bench>`
+    panic!("Enable the criterion_bench feature to run benchmarks");
+}

